### ISSUE TITLE
Odd complex-number shrinking when min/max magnitude

### ISSUE THREAD

**Issue by ghost** [HypothesisWorks/hypothesis#1212]
Observed in the PR #1154, but there decided to convert to an issue rather than block the merge.  The shrinking of complex numbers in the presence of a `min`/`max_magnitude` constraint seems odd. Probably best demonstrated by examples:
```python
def test_minimal_min_magnitude_dyadic():
    assert minimal(
        complex_numbers(min_magnitude=0.5),
        lambda x: True
    ) == 1
# AssertionError: assert (0.5+0j) == 1

def test_minimal_min_magnitude_messy():
    assert minimal(
        complex_numbers(min_magnitude=0.501884271),
        lambda x: True
    ) == 1
# AssertionError: assert (0.501884271+0j) == 1

def test_minimal_min_magnitude_ovelapped():
    assert minimal(
        complex_numbers(min_magnitude=0.5),
        lambda x: x.real > 0.5
    ) == 1
# AssertionError: assert (1.5+0j) == 1

def test_minimal_min_magnitude_nonzero():
    assert minimal(
        complex_numbers(min_magnitude=0.5),
        lambda x: x.real != 0.0
    ) == 1
# AssertionError: assert (0.5+0j) == 1
```

**Comment by Zac-HD** [comment]
This may be related to #909 - or at least it has similar symptoms.

**Comment by Zac-HD** [comment]
Uh-oh.  The root cause is actually that `floats()` only uses the documented shrink order if there are no bounds.  Fixing this will be... complicated, much worse than the corresponding problem with `integers()` (fixed in #1616).  I'll write up a detailed issue to replace this one, including some solutions for discussion.

**Comment by Zac-HD** [comment]
Superseded by #1704.
